app/queries/digest_items_query.rb

```diff
@@ -20,54 +28,35 @@ def call
-      memo[id] ||= {
-        subscriber_list_title: record[:subscriber_list_title],
-        subscriber_list_url: record[:subscriber_list_url],
-      }
-      memo[id][:content_changes] = Array(memo[id][:content_changes]) << record
-    end
-
-    result_data = messages.each_with_object(result_data) do |record, memo|
-      id = record[:subscription_id]
-      memo[id] ||= {
-        subscriber_list_title: record[:subscriber_list_title],
-        subscriber_list_url: record[:subscriber_list_url],
-      }
-      memo[id][:messages] = Array(memo[id][:messages]) << record
-    end
-
-    result_data.map do |key, value|
-      content = value.fetch(:content_changes, []) + value.fetch(:messages, [])
-      Result.new(key, value[:subscriber_list_title], value[:subscriber_list_url], content.sort_by(&:created_at))
-    end
-  end
-
-  def fetch_content_changes
-    ContentChange
-      .select("content_changes.*", "subscriptions.id AS subscription_id", "subscriber_lists.title AS subscriber_list_title", "subscriber_lists.url AS subscriber_list_url")
-      .joins(matched_content_changes: { subscriber_list: { subscriptions: :subscriber } })
-      .where(subscribers: { id: subscriber.id })
-      .where(subscriptions: { frequency: Subscription.frequencies[digest_run.range] })
+  def content_changes
+    @content_changes ||= ContentChange
+      .select("content_changes.*, subscriptions.id AS subscription_id")
+      .joins(matched_content_changes: { subscriber_list: :subscriptions })
+      .where(subscriptions: { id: subscriptions })
       .where("content_changes.created_at >= ?", digest_run.starts_at)
       .where("content_changes.created_at < ?", digest_run.ends_at)
-      .merge(Subscription.active)
-      .order("subscriber_list_title ASC", "subscriber_list_url ASC", "content_changes.created_at ASC")
+      .order("subscriber_lists.title ASC", "content_changes.created_at ASC")
       .uniq(&:content_id)
+      .group_by(&:subscription_id)
   end
 
-  def fetch_messages
-    Message
-      .select("messages.*", "subscriptions.id AS subscription_id", "subscriber_lists.title AS subscriber_list_title", "subscriber_lists.url AS subscriber_list_url")
-      .joins(matched_messages: { subscriber_list: { subscriptions: :subscriber } })
-      .where(subscribers: { id: subscriber.id })
-      .where(subscriptions: { frequency: Subscription.frequencies[digest_run.range] })
+  def messages
+    @messages ||= Message
+      .select("messages.*, subscriptions.id AS subscription_id")
+      .joins(matched_messages: { subscriber_list: :subscriptions })
+      .where(subscriptions: { id: subscriptions })
       .where("messages.created_at >= ?", digest_run.starts_at)
       .where("messages.created_at < ?", digest_run.ends_at)
-      .merge(Subscription.active)
-      .order("subscriber_list_title ASC", "subscriber_list_url ASC", "messages.created_at ASC")
+      .order("subscriber_lists.title ASC", "messages.created_at ASC")
       .uniq(&:id)
```

Comment:
Yes, that was confusing - that and the `order` clauses. I've added a commit for the latter, as I don't think they're necessary. I'm also confused about why we need to remove content changes for the same Content ID, as presumably they all reflect distinct changes the subscriber may want to learn about.

For messages, it's even more confusing: there's [a test](https://github.com/alphagov/email-alert-api/blob/f1687ad93e261a1ec4368072407b5221d68fcb7a/spec/queries/digest_items_query_spec.rb#L153-L166) that specifically requires that `uniq(&:id)` call to be there. Even when digests were a single email, it still doesn't make sense to only show the change under one of the lists - we're effectively doing the same for content changes with the `uniq` call there, too.

I'm concerned this might be a bit of a rabbit hole, though. While it seems like both of those `uniq` calls can be removed, I'd prefer to do that in a follow-up PR, if we do it at all.

Reply:
Sure, I recall the `content_id` one being a product decision that it was confusing that you could have conflicting updates to the same content item in a digest, however I doubt that is a documented product decision and seems rather heavily caveated where things could be confusing.

I can't help but think the id one on Message is just a mistake though and that the test you'd referenced would pass without it - given ActiveRecord won't return the same model twice. Not too concerned if you want to defer it though.

---

Nice one on the ordering. I imagine prior to messages we didn't need to do any ordering in Ruby but now there are two things combined.